api/src/main/java/org/openmrs/tag/api/impl/TagServiceImpl.java

```diff
@@ -58,7 +58,11 @@ public Tag saveTag(Tag tag) {
 		if (getObject(toClass(tag.getObjectType()), tag.getObjectUuid()) == null) {
 			log.warn("Object does not exist in the the database");
 		}
-		return dao.saveTag(tag);
+		if (hasTag(getObject(toClass(tag.getObjectType()), tag.getObjectUuid()), tag.getTag())) {
+			throw new IllegalArgumentException("Duplicate Tag");
```

Comment:
I hope this exception doesn't bubble to the UI, the UI should silently just ignore it if an object already has the tag

Reply:
It does. I thought you wanted a warning in the UI. If i log a warning instead of throwing an exception, the ui will ignore it

Reply:
The UI should do nothing since the duplicate tag is ignored on the sever side, you should catch the error in the rest resource and do nothing.

Reply:
Are you still working on this @jtatia ?

Reply:
Yes @wluyima . I have exams till the 28th. I'll get back to working on the module after that

Reply:
Did you address this?

Reply:
@wluyima . Since i am getting back to this after quite some while I just wanted to confirm something, there are 2 approaches for handling duplicate tags. one is to handle them on the client side itself by checking the list of current tags. The other approach is to silently handle the exception on the server side. This seems to be handling both. It handles duplicates by keeping a count and flag variable, which checks if the tag being added already exists. In case the request gets sent to the server side even though its a duplicate, it throws an exception (not required so I'll remove this). 
Are we going with handling things on the server side only ?

Reply:
After all this time, I've lost track of this, I believe catching the error on the client side is to be able to show the user a specific and informative error message which you wouldn't be able to do if you depended on the server side validation since the error you would get back is generic and isn't meant to be displayed to a user for security purposes.